apps/server/src/modules/authentication/uc/account.uc.ts

```diff
@@ -15,45 +20,125 @@ export class AccountUc {
+		}
 
-	private async updatePassword(account: Account, password: string) {
-		account.password = await bcrypt.hash(password, 10);
-		await this.accountRepo.update(account);
-	}
+		if (!params.passwordOld || !account.password || !(await this.checkPassword(params.passwordOld, account.password))) {
+			throw new Error('Dein Passwort ist nicht korrekt!');
+		}
 
-	private hasPermissionsToChangePassword(currentUser: User, targetUser: User) {
-		if (this.hasRole(currentUser, 'superhero')) {
-			return true;
+		let user: User;
+		try {
+			user = await this.userRepo.findById(currentUser, true);
+		} catch (err) {
+			throw new EntityNotFoundError('User');
 		}
-		if (!(currentUser.school.id === targetUser.school.id)) {
-			return false;
+
+		let updateUser = false;
+		let updateAccount = false;
+		if (params.passwordNew) {
+			account.password = await this.calcPasswordHash(params.passwordNew);
+			updateAccount = true;
 		}
-		if (this.isDemoUser(currentUser)) {
-			return false;
+		// TODO Check with BC-1471 - This might be removed from the account page
+		if (params.language && user.language !== params.language) {
+			// TODO Check if there is an enum or some other fixed values here?
+			user.language = params.language;
+			updateUser = true;
+		}
+		if (params.email && user.email !== params.email) {
+			const newMail = params.email.toLowerCase();
+			await this.checkUniqueEmail(account, user, newMail);
+			user.email = newMail;
+			account.username = newMail;
+			updateUser = true;
+			updateAccount = true;
 		}
 
-		const permissionsToCheck: string[] = [];
-		if (this.hasRole(targetUser, 'student')) {
-			permissionsToCheck.push('STUDENT_EDIT');
+		if (params.firstName && user.firstName !== params.firstName) {
+			if (!this.hasPermissionsToChangeOwnName(user)) {
+				throw new InvalidOperationError('No permission to change first name');
+			}
+			user.firstName = params.firstName;
+			updateUser = true;
 		}
-		if (this.hasRole(targetUser, 'teacher')) {
-			permissionsToCheck.push('TEACHER_EDIT');
+		if (params.lastName && user.lastName !== params.lastName) {
+			if (!this.hasPermissionsToChangeOwnName(user)) {
+				throw new InvalidOperationError('No permission to change last name');
+			}
+			user.lastName = params.lastName;
+			updateUser = true;
 		}
-		if (permissionsToCheck.length === 0) {
-			// target user is neither student nor teacher. Undefined what to do
-			return false;
+
+		if (updateUser) {
+			try {
+				await this.userRepo.update(user);
+			} catch (err) {
+				throw new EntityNotFoundError(User.name);
+			}
+		}
+		if (updateAccount) {
+			try {
+				await this.accountRepo.update(account);
+			} catch (err) {
+				throw new EntityNotFoundError(Account.name);
+			}
 		}
+	}
 
-		return this.permissionService.hasUserAllSchoolPermissions(currentUser, permissionsToCheck);
+	async changeMyTemporaryPassword(userId: EntityId, password: string, confirmPassword: string): Promise<void> {
+		if (password !== confirmPassword) {
```

Comment:
The use case is to set a new password just after I did my first login or a login after a force password reset. In both cases I, the user, does not have a 'real password'. Hence it is not checked. This method is for setting my, the users, 'real password'. The client does not require to give the old password. 

The business logic demands that we have 2 strings that match each other (in both cases either first login or force reset). Hence we are validating that exact logic in the server. The client must not know about the details although it might validate it on its own of course.